### ISSUE TITLE
Async and addFetchCondition / queueadd

### ISSUE THREAD

**Issue by maxcorbeau** [cgiffard/node-simplecrawler#345]
I am trying to use some `aync` code in `addFetchCondition` and use the result of that code as the result of the fetch condition:

```
// Only fetch URLs if we haven't in the past (stored in Mongo)
crawler.addFetchCondition(function(queueItem, referrerQueueItem) {
    var fetch = false;
    // Check if URL already stored in Mongo
    mongodb.collection('urls').findOne({
        url: queueItem.url
    }, function(err, result) {
        fetch = (result)? false: true; // use this as condition 
    });
    // mongodb call is async so get here before mongodb callback
});
```

Obviously the above code doesn't work as `mongodb` calls are `async` so we exit `addFetchCondition` before we hear back from `mongo`.

Looking at the doc I see a method  (`var continue = this.wait(); | continue();`) to wait within event listeners:
https://github.com/cgiffard/node-simplecrawler#waiting-for-asynchronous-event-listeners

But this method doesn't work within `addFetchCondition` (the `this` object doesn't exist, and trying the `on('addFetchCondition','function'` doesn't work.

**How can I run `async` code within `addFetchCondition` and wait for it?**

Alternative seems to be using `queueadd` which is an event listener but `this.wait()` logic doesn't work

```
    crawler.on('queueadd', function(queueItem, data, res) {
        console.log('queueadd',queueItem.url);
        var resume = this.wait();
        setTimeout(function(){
            resume();
        },10000);
    });
```

**How can I use `queueadd` to wait for async code and remove item from queue?**

**Comment by edgahan** [comment]
Hi, 

I'm trying to do the same as you, but with Elasticsearch -
Not sure if this helps, but from quickly reading the source and trying a few different ways of accomplishing this (looked at filtering`discoverResources`, etc) - the (probably not amazing) way that seems to be working for me is:

```
    crawler.addFetchCondition(function(queueItem, referrerQueueItem) {
        if (queueItem.esChecked) return true;

        elasticsearch.checkExists(queueItem.uriPath, function(exists) {
          queueItem.esChecked = true;
          if (!exists) {
            console.log("queueItem does not exist, queuing", queueItem.uriPath)
            crawler.queueURL(queueItem, referrerQueueItem);
          } else {
            console.log("queueItem already exists in elasticsearch", queueItem.uriPath);
          }
        })
        return false;
    });
```

So basically I make `addFetchCondition` always return false, I do the async check with elasticsearch, if not present in ES then I set a property on the queueItem and re-queue it. Once re-queued it has the `esChecked` flag set, so it passes and you can see the item being set with the `.on('queueadd', function(queueItem)` event

**Comment by fredrikekelund** [comment]
Hi @maxcorbeau, sorry for the very slow response here. This is a tricky one - you're quite right that simplecrawler's API wasn't designed to allow asynchronous fetch conditions. There are of course performance drawbacks of doing async operations in fetch conditions because of how often they are executed, but I totally understand the need for it and would honestly say that this is something we should probably look into adding. If nothing else because of the fact that the `FetchQueue#add` method is asynchronous for the purposes of allowing these kinds of database lookups. 

So the "right" way to do this with the current version would be to monkey patch the `FetchQueue#add` method.

But I would also like to go ahead and add the ability to have asynchronous fetch conditions in a future version. This is very much inline with our efforts to update the FetchQueue methods that led up to the 1.0 release. Thoughts, @cgiffard?

**Comment by cgiffard** [comment]
I'm inclined to make a breaking change — maybe even a 2.0 — and add async fetch conditions as the default — and only method. The interface would be exactly the same, except it'd provide a callback too.

**Comment by cgiffard** [comment]
Definitely for DB backed systems it's important.

**Comment by fredrikekelund** [comment]
Couldn't we support both sync and async interfaces by simply checking the number of arguments that the fetch condition accepts? I agree that when we release a new major version, it makes sense to make all fetch conditions asynchronous, but feels like we could gather up a few more breaking changes before releasing a 2.0

**Comment by cgiffard** [comment]
Yep, good point. I'd still like to deprecate the synchronous version, (maybe then remove it in 2.0) in the interests of reducing complexity, both in the API and code.

**Comment by maxcorbeau** [comment]
@fredrikekelund: please don't apologize for delay. I'm very greatful you guys find time to develop and maintain `simple-crawler` and make it open source, and delighted to see aync conditions are making their way into the roadmap :)

**Comment by fredrikekelund** [comment]
I've opened a PR that addresses this and adds the ability to have async fetch conditions. Will cut a new release once we get that merged!

**Comment by fredrikekelund** [comment]
It took some extra time to churn through the PR, because I realized that we would want to support async download conditions as well. It's been merged now however, and simplecrawler 1.1.0 has been published to npm!

**Comment by AlexoARG** [comment]
@edgahan I did what you recommend. Now I have the problem that those URLs that I add manually do not crawl them, I did exactly what you mention. I do the chek in the queueadd function and it apparently adds them, but it does not analyze them

I'm forgetting something?
